crypto/ocsp/ocsp_integration_test.cc

```diff
@@ -0,0 +1,212 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0 OR ISC
+
+#include <gtest/gtest.h>
+#include <ctime>
+
+#include <openssl/ocsp.h>
+#include <openssl/pem.h>
+#include <openssl/ssl.h>
+
+#include "../internal.h"
+#include "../../tool/transport_common.h"
+#include "internal.h"
+
+#define OCSP_VERIFYSTATUS_SUCCESS 1
+#define OCSP_VERIFYSTATUS_ERROR 0
+#define OCSP_VERIFYSTATUS_FATALERROR -1
+
+std::string GetTestData(const char *path);
+
+static OCSP_RESPONSE *GetOCSPResponse(const char *ocsp_responder_host,
+                            X509 *certificate,
+                            OCSP_REQUEST *request) {
+  // Connect to OCSP Host. Connect to the defined OCSP responder if specified.
+  // Otherwise, we connect to the OCSP responder specified in the cert.
+  char *host = nullptr;
+  bssl::UniquePtr<BIO> http_bio;
+  if (ocsp_responder_host) {
+    http_bio = bssl::UniquePtr<BIO>(BIO_new_connect(ocsp_responder_host));
+  } else {
+    bssl::UniquePtr<STACK_OF(OPENSSL_STRING)> responder(
+        X509_get1_ocsp(certificate));
+    char *path, *port;
+    int is_ssl;
+    EXPECT_TRUE(OCSP_parse_url(sk_OPENSSL_STRING_value(responder.get(), 0),
+                               &host, &port, &path, &is_ssl));
+    http_bio = bssl::UniquePtr<BIO>(BIO_new_connect(host));
+    OPENSSL_free(path);
+    OPENSSL_free(port);
+  }
+  EXPECT_TRUE(http_bio);
+  BIO_set_conn_port(http_bio.get(), "80");
+  BIO_set_retry_read(http_bio.get());
+  EXPECT_TRUE(BIO_do_connect(http_bio.get()));
+
+  // Set up an |OCSP_REQ_CTX| to be sent out.
+  bssl::UniquePtr<OCSP_REQ_CTX> ocsp_ctx(
+      OCSP_sendreq_new(http_bio.get(), "/", nullptr, -1));
+  OCSP_REQ_CTX_add1_header(
+      ocsp_ctx.get(), "Host",
+      ocsp_responder_host ? ocsp_responder_host : host);
+  OCSP_REQ_CTX_set1_req(ocsp_ctx.get(), request);
+
+  // Try connecting to the OCSP responder endpoint with a timeout of 3 seconds.
+  // Sends out an OCSP request and expects an OCSP response.
+  OCSP_RESPONSE *resp = nullptr;
+  time_t start_time = time(nullptr);
+  double timeout = 3;
+  int rv;
+  do {
+    rv = OCSP_sendreq_nbio(&resp, ocsp_ctx.get());
+  } while ((rv == -1) && difftime(time(nullptr), start_time) < timeout);
+
+  OPENSSL_free(host);
+  return resp;
+}
+
+static void ValidateOCSPResponse(OCSP_RESPONSE *response,
+                                 bool authorized_responder,
+                                 STACK_OF(X509) *chain,
+                                 OCSP_CERTID *cert_id,
+                                 int expected_verify_status,
+                                 int expected_cert_status) {
+  ASSERT_EQ(OCSP_response_status(response),
+            authorized_responder ? OCSP_RESPONSE_STATUS_SUCCESSFUL
+                                 : OCSP_RESPONSE_STATUS_UNAUTHORIZED);
+  if (!authorized_responder) {
+    return;
+  }
```

Comment:
Is it possible/interesting to move this down below after the OCSP_response_get1_basic and if it's not a authorized responder we expect there is garbage or nothing in the response and basic_response would be null? Basically a very tiny fuzz test that if someone connects to a bad endpoint our code fails and fails gracefully.

Reply:
Good point, I think that totally makes sense. I added an assert after the basic response parsing, it should always be false in these cases.